openedx_external_enrollments/external_enrollments.py

```diff
@@ -0,0 +1,424 @@
+        """
+        """
+        user, _ = get_user(email=data.get("user_email"))
+        return {
+            "user": user.username,
+            "is_active": True,
+            "mode": course_settings.get(
+                "external_enrollment_mode_override",
+                data.get("course_mode")
+            ),
+            "course_details": {
+                "course_id": course_settings.get("external_course_run_id")
+            }
+        }
+
+    def _get_enrollment_url(self, course_settings):
+        """
+        """
+        return course_settings.get("external_enrollment_api_url")
+
+
+class SalesforceEnrollment(BaseExternalEnrollment):
+
+    def __str__(self):
+        return "salesforce"
+
+    @staticmethod
+    def _get_course(course_id):
+        """
+        Return a course object.
+        """
+        if not course_id:
+            return None
+
+        course_key = CourseKey.from_string(course_id)
+        course = get_course_by_id(course_key)
+        return course
+
+    def _get_enrollment_headers(self):
+        token = self._get_auth_token()
+        return {
+            "Content-Type": "application/json",
+            "Authorization": "{} {}".format(
+                token.get("token_type"),
+                token.get("access_token"),
+            )
+        }
+
+    @staticmethod
+    def _get_auth_token():
+        """
+
+        :return:
+        """
+        request_params = {
+            'client_id': settings.SALESFORCE_API_CLIENT_ID,
+            'client_secret': settings.SALESFORCE_API_CLIENT_SECRET,
+            'username': settings.SALESFORCE_API_USERNAME,
+            'password': settings.SALESFORCE_API_PASSWORD,
+            'grant_type': 'password',
+        }
+        client = BackendApplicationClient(**request_params)
+        oauth = OAuth2Session(client=client)
+        oauth.params = request_params
+        token = oauth.fetch_token(
+            token_url=settings.SALESFORCE_API_TOKEN_URL,
+        )
+
+        return token
+
+    @staticmethod
+    def _get_openedx_user(data):
+        """
+
+        :param data:
+        :return:
+        """
+        user = {}
+        order_lines = data.get("supported_lines")
+        if order_lines:
+            try:
+                email = order_lines[0].get("user_email")
+                openedx_user, openedx_profile = get_user(email=email)
+                # TODO do not force logic assuming names with 2 words
+                first_name, last_name = openedx_profile.name.split(" ", 1)
+                user["FirstName"] = first_name.strip(" ")
+                user["LastName"] = last_name.strip(" ")
+                user["Email"] = email
+            except Exception:
+                pass
+
+        return user
+
+    def _get_salesforce_data(self, data):
+        """
+
+        :param data:
+        :return:
+        """
+        salesforce_data = {}
+        order_lines = data.get("supported_lines")
```

Comment:
Object of type 'NoneType' has no len() if it takes the default value. @andrey-canon

Reply:
out of the scope of this pr